File: modules/core-api/src/main/java/com/enonic/xp/core/content/page/region/Descriptor.java
package com.enonic.xp.core.content.page.region;

import com.google.common.base.Preconditions;

import com.enonic.xp.core.content.page.DescriptorKey;
import com.enonic.xp.core.form.Form;
import com.enonic.xp.core.resource.ResourceKey;

public abstract class Descriptor<KEY extends DescriptorKey>
{
    private final KEY key;

    private final String name;

    private final String displayName;

    private final Form config;

    protected Descriptor( final BaseDescriptorBuilder builder )
    {
        Preconditions.checkNotNull( builder.key, "key cannot be null" );
        Preconditions.checkNotNull( builder.name, "name cannot be null" );
        Preconditions.checkNotNull( builder.config, "config cannot be null" );
        //noinspection unchecked
        this.key = (KEY) builder.key;
        this.name = builder.name;
        this.displayName = builder.displayName;
        this.config = builder.config != null ? builder.config : Form.newForm().build();
    }

    public KEY getKey()
    {
        return key;
    }

    public String getName()
    {
        return name;
    }

    public String getDisplayName()
    {
        return displayName;
    }

    public Form getConfig()
    {
        return config;
    }

    public abstract ResourceKey getComponentPath();

    public abstract static class BaseDescriptorBuilder<T extends Descriptor.BaseDescriptorBuilder, KEY>
    {
        protected KEY key;

        protected String name;

        protected String displayName;

        protected Form config;

        public T key( final KEY key )
        {
            this.key = key;
            return typecastToBuilder( this );
        }

        public T name( final String name )
        {
            this.name = name;
            return typecastToBuilder( this );
        }

        public T displayName( final String displayName )
        {
            this.displayName = displayName;
            return typecastToBuilder( this );
        }

        public T config( final Form value )
        {
            this.config = value;
            return typecastToBuilder( this );
        }

        @SuppressWarnings("unchecked")
        private T typecastToBuilder( final BaseDescriptorBuilder object )
        {
            return (T) object;
        }

    }
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/content/page/Descriptor.ts
module api.content.page {

    export class Descriptor implements api.Cloneable {

        private key: DescriptorKey;

        private name: DescriptorName;

        private displayName: string;

        private config: api.form.Form;

        constructor(builder: DescriptorBuilder) {
            this.name = builder.name;
            this.key = builder.key;
            this.displayName = builder.displayName;
            this.config = builder.config;
        }

        getKey(): DescriptorKey {
            return this.key;
        }

        getName(): DescriptorName {
            return this.name;
        }

        getDisplayName(): string {
            return this.displayName;
        }

        getConfig(): api.form.Form {
            return this.config;
        }

        clone(): Descriptor {
            throw new Error("Must be implemented in inheritor");
        }
    }

    export class DescriptorBuilder {

        key: DescriptorKey;

        name: DescriptorName;

        displayName: string;

        config: api.form.Form;

        constructor(source?: Descriptor) {
            if (source) {
                this.key = source.getKey();
                this.name = source.getName();
                this.displayName = source.getDisplayName();
                this.config = source.getConfig();
            }
        }
    }
}